pkg/forward/forward.go

```diff
@@ -478,53 +478,73 @@ func (isdf *InterStepDataForward) writeToBuffers(ctx context.Context, messageToS
 
 // writeToBuffer forwards an array of messages to a single buffer and is a blocking call or until shutdown has been initiated.
 func (isdf *InterStepDataForward) writeToBuffer(ctx context.Context, toBuffer isb.BufferWriter, messages []isb.Message) (writeOffsets []isb.Offset, err error) {
-	var totalBytes float64
-	for _, m := range messages {
-		totalBytes += float64(len(m.Payload))
-	}
-	writeOffsets = make([]isb.Offset, 0, len(messages))
-retry:
-	needRetry := false
+	var (
+		totalCount int
+		writeCount int
+		writeBytes float64
+		dropBytes  float64
+	)
+	totalCount = len(messages)
+	writeOffsets = make([]isb.Offset, 0, totalCount)
+
 	for {
 		_writeOffsets, errs := toBuffer.Write(ctx, messages)
 		// Note: this is an unwanted memory allocation during a happy path. We want only minimal allocation since using failedMessages is an unlikely path.
-		var failedMessages []isb.Message
-		for idx := range messages {
-			// use the messages index to index error, there is a 1:1 mapping
-			err := errs[idx]
-			// ATM there are no user defined errors during write, all are InternalErrors, and they require retry.
-			if err != nil {
-				needRetry = true
-				// we retry only failed messages
-				failedMessages = append(failedMessages, messages[idx])
-				writeMessagesError.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName, "buffer": toBuffer.GetName()}).Inc()
-				// a shutdown can break the blocking loop caused due to InternalErr
-				if ok, _ := isdf.IsShuttingDown(); ok {
-					err := fmt.Errorf("writeToBuffer failed, Stop called while stuck on an internal error with failed messages:%d, %v", len(failedMessages), errs)
-					platformError.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName}).Inc()
-					return writeOffsets, err
+		failedMessages := make([]isb.Message, 0, len(messages))
```

Comment:
This is initializing the length and capacity?

Reply:
Addressed.

Yes. I think if retrying is frequent , pre-allocate memory could be more efficient because it help us avoid resizing the slice repeatedly and minimize the number of allocations and copies needed during the append operations.

But I think for now we are not expecting unhappy path, I changed it back.